src/main/java/org/opensearch/security/http/HTTPClientCertAuthenticator.java

```diff
@@ -54,66 +63,126 @@ public class HTTPClientCertAuthenticator implements HTTPAuthenticator {
-                    if (usernames.isEmpty() == false) {
-                        username = usernames.get(0);
-                    }
+            Pattern pattern = null;
+            if (!Strings.isNullOrEmpty(regex)) {
+                try {
+                    pattern = Pattern.compile(regex, Pattern.CASE_INSENSITIVE | Pattern.UNICODE_CASE);
+                } catch (Exception e) {
+                    log.warn("Invalid regex in attribute '{}': {}", raw, e.toString());
+                    return null;
                 }
+            }
 
-                if (rolesAttribute != null && rolesAttribute.length() > 0) {
-                    final List<String> roles = getDnAttribute(rfc2253dn, rolesAttribute);
-                    if (roles.isEmpty() == false) {
-                        backendRoles = roles.toArray(new String[0]);
-                    }
-                }
+            return ParsedAttribute.san(new ParsedSAN(sanTypeInt, pattern));
+        }
 
-                return new AuthCredentials(username, backendRoles).markComplete();
-            } catch (InvalidNameException e) {
-                log.error("Client cert had no properly formed DN (was: {})", principal);
-                return null;
-            }
+        // DN form: either "dn:cn" or just "cn"
+        final String dnAttr = s.regionMatches(true, 0, "dn:", 0, 3) ? s.substring(3) : s;
+        return ParsedAttribute.dn(dnAttr);
+    }
 
-        } else {
+    public HTTPClientCertAuthenticator(final Settings settings, final Path configPath) {
+        this.settings = settings;
+        this.skipUsersMatcher = WildcardMatcher.from(settings.getAsList(OPENDISTRO_SECURITY_SSL_SKIP_USERS));
+        this.parsedUsernameAttr = parseAttributeSetting(settings.get("username_attribute"));
+        this.parsedRolesAttr = parseAttributeSetting(settings.get("roles_attribute"));
+    }
+
+    @Override
+    public AuthCredentials extractCredentials(final SecurityRequest request, final ThreadContext threadContext) {
+
+        final String principal = threadContext.getTransient(ConfigConstants.OPENDISTRO_SECURITY_SSL_PRINCIPAL);
+
+        if (Strings.isNullOrEmpty(principal)) {
             log.trace("No CLIENT CERT, send 401");
             return null;
         }
+        if (skipUsersMatcher.test(principal)) {
+            log.debug("Skipped user client cert authentication of user {} as its in skip_users list ", principal);
+            return null;
+        }
+
+        try {
+            final String username = extractUsername(threadContext, principal);
+            final String[] roles = extractRoles(threadContext, principal);
+            return new AuthCredentials(username, roles).markComplete();
+        } catch (InvalidNameException e) {
+            log.error("Client cert had no properly formed DN");
+            log.debug("Client cert had no properly formed DN (was: {})", principal);
```

Comment:
nit: this would be repetitive with logs at debug level. Considering only logging the DN with this log line. i.e. `log.debug("DN (was: {})", principal);`